### PR TITLE
Staggered flipping now works as expected, as well, Hexagonal rotation now works in 60 degree intervals.

### PR DESCRIPTION
**Pull request by Bdtrotte**
Addresses #1476. Adds a hex class, which stores info of the cube coordinates of a hex tile, and implements methods for rotation, and converting to staggered coordinates.

### CODE REVIEW

**Comment by Bdtrotte** [comment]
I think I've done something wrong, as it's showing commits far beyond the context of this pr. Ill work to fix this...

**Comment by Bdtrotte** [comment]
Alright, should be up to date now

**Comment by bjorn** [comment]
> Alright, should be up to date now

Well, you merged my branch with yours, but you should rather rebase your work and in the process you can remove all the commits that have already been merged (and squashed as cdc71f4bf06c2ce9fb754887a8756f3686b18ea2).

File: src/libtiled/tilelayer.h [hunk 1]
```diff
@@ -266,7 +268,7 @@ class TILEDSHARED_EXPORT TileLayer : public Layer
      * are rotated within the layer, and the tiles themselves are rotated. The
      * dimensions of the tile layer are swapped.
      */
-    void rotateHexagonal(RotateDirection direction);
+    void rotateHexagonal(RotateDirection direction, Tiled::Map::StaggerIndex staggerIndex, Tiled::Map::StaggerAxis staggerAxis);
```
Comment (bjorn) [inline src/libtiled/tilelayer.h]: You're in the `Tiled` namespace here, so you don't need the `Tiled::` qualification.

File: src/tiled/tilestamp.cpp [hunk 2]
```diff
@@ -232,6 +232,22 @@ TileStamp TileStamp::flipped(FlipDirection direction) const
...
         TileLayer *layer = variation.tileLayer();
+
+        Map::StaggerIndex staggerIndex = variation.map->staggerIndex();
+        Map::StaggerAxis staggerAxis = variation.map->staggerAxis();
+
+        if (variation.map->orientation() == Map::Staggered || variation.map->orientation() == Map::Hexagonal) {
+            if (staggerAxis == Map::StaggerY)
+            {
+                if ((direction == FlipVertically && !(layer->height() & 1)) || direction == FlipHorizontally)
+                    variation.map->setStaggerIndex(static_cast<Map::StaggerIndex>((staggerIndex + 1) & 1));
```
Comment (bjorn) [inline src/tiled/tilestamp.cpp]: A little shorter would be:
```c++
variation.map->setStaggerIndex(static_cast<Map::StaggerIndex>(1 - staggerIndex));
```
Though for readability I might try to use:
```c++
static Map::StaggerIndex invert(Map::StaggerIndex staggerIndex)
{
    return static_cast<Map::StaggerIndex>(1 - staggerIndex);
}
```

File: src/tiled/tilestamp.cpp [hunk 3]
```diff
@@ -232,6 +232,22 @@ TileStamp TileStamp::flipped(FlipDirection direction) const
 
     for (const TileStampVariation &variation : flipped.variations()) {
         TileLayer *layer = variation.tileLayer();
+
+        Map::StaggerIndex staggerIndex = variation.map->staggerIndex();
+        Map::StaggerAxis staggerAxis = variation.map->staggerAxis();
+
+        if (variation.map->orientation() == Map::Staggered || variation.map->orientation() == Map::Hexagonal) {
```
Comment (bjorn) [inline src/tiled/tilestamp.cpp]: Remember that we added `map->isStaggered()`.

File: src/libtiled/hex.cpp [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+Hex::Hex(int x, int y, int z):
+    mX(x),
+    mY(y),
+    mZ(z) { }
+
+Hex::Hex(QPoint point,
+         Map::StaggerIndex staggerIndex,
+         Map::StaggerAxis staggerAxis)
+{
+    Hex(point.x(),point.y(),staggerIndex,staggerAxis);
```
Comment (Bdtrotte) [inline src/libtiled/hex.cpp]: Thanks for this. I was thinking logically what I was doing wasn't quite right, but thought I would get an error if it was wrong. But I see now exactly what it was doing. 

> instead you could also define a Hex::setStaggered(int col, int row, ...)

Good advice! Implementing this way.

File: src/tiled/tilestamp.cpp (hunk 3 above)
Comment (Bdtrotte) [inline src/tiled/tilestamp.cpp]: I forgot to use the function I added... Resolved now!

**Comment by Bdtrotte** [comment]
> Well, you merged my branch with yours, but you should rather rebase your work and in the process you can remove all the commits that have already been merged (and squashed as cdc71f4).

Would it be best to close this pr and start a new one with everything cleaned up? Or just continue working in this one?

**Comment by Bdtrotte** [comment]
Updated all items from the review. I absolutely appreciate good style, just have to get used to being consistently good! As well constantly learning a lot about c++, so hopefully less coding errors as we go.

**Comment by bjorn** [comment]
> Would it be best to close this pr and start a new one with everything cleaned up? Or just continue working in this one?

There is no reason to close this PR, please keep updating this one.

**Comment by Bdtrotte** [comment]
I believe I have fixed all the style issues. Is there anything else you think should be changed within the code?

File: src/libtiled/tilelayer.cpp [hunk 5]
```diff
@@ -360,16 +372,34 @@ void TileLayer::rotateHexagonal(RotateDirection direction)
...
 
     mWidth = newWidth;
     mHeight = newHeight;
     mGrid = newGrid;
+
+    QRect filledRect = region().boundingRect();
+
+    if (staggerAxis == Map::StaggerY) {
+        if (filledRect.y() & 1)
+            filledRect.adjust(0, -1, 0, 0);
```
Comment (bjorn) [inline src/libtiled/tilelayer.cpp]: Would it be possible to change the stagger index instead of keeping an empty row of data around?

File: src/libtiled/tilelayer.cpp [hunk 6]
```diff
@@ -312,12 +312,24 @@ void TileLayer::rotate(RotateDirection direction)
     mGrid = newGrid;
 }
 
-void TileLayer::rotateHexagonal(RotateDirection direction)
+void TileLayer::rotateHexagonal(RotateDirection direction, Map::StaggerIndex staggerIndex, Map::StaggerAxis staggerAxis)
 {
-    int newWidth = mHeight;
-    int newHeight = mWidth;
+    Hex botRight(mWidth, mHeight, staggerIndex, staggerAxis);
```
Comment (bjorn) [inline src/libtiled/tilelayer.cpp]: Naming style: Please just use `bottomRight`.

File: src/libtiled/tilelayer.cpp [hunk 7]
```diff
@@ -312,12 +312,24 @@ void TileLayer::rotate(RotateDirection direction)
...
+    Hex center(mWidth / 2, mHeight / 2, staggerIndex, staggerAxis);
+
+    botRight -= center;
+    topRight -= center;
+
+    botRight.rotate(RotateRight);
+    topRight.rotate(RotateRight);
+
+    int newWidth = topRight.toStaggered(staggerIndex, staggerAxis).x() * 2 + 2;
+    int newHeight = botRight.toStaggered(staggerIndex, staggerAxis).y() * 2 + 2;
```
Comment (bjorn) [inline src/libtiled/tilelayer.cpp]: It is not immediately obvious to me why this works (like, why the new area is both minimal but also large enough). Could you share a few words about it?

File: src/libtiled/hex.h [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+    void setZ(int z) { mZ = z; }
+
+    void setStaggered(int col, int row,
+                      Map::StaggerIndex staggerIndex,
+                      Map::StaggerAxis staggerAxis);
+
+    Hex operator + (Hex h) const;
+    Hex operator -(Hex h) const;
+    Hex operator *(float f) const;
+    Hex operator /(float f) const;
```
Comment (bjorn) [inline src/libtiled/hex.h]: You're not actually using the `*` and `/` operators and using them would risk violating the rule that hex cube coordinates always adhere to `x + y + z = 0` (due to integer truncation). I would suggest you leave those out.

File: src/libtiled/hex.h [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+
+    void setX(int x) { mX = x; }
+    void setY(int y) { mY = y; }
+    void setZ(int z) { mZ = z; }
+
+    void setStaggered(int col, int row,
+                      Map::StaggerIndex staggerIndex,
+                      Map::StaggerAxis staggerAxis);
+
+    Hex operator + (Hex h) const;
```
Comment (bjorn) [inline src/libtiled/hex.h]: Stray space after the `+` (it'd be fine, but you don't have it on any of the other operators).

File: src/libtiled/hex.h [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+
+#pragma once
+
+#include "tiled.h"
+#include "map.h"
+#include <QPoint>
+
+namespace Tiled {
+
+class Hex
```
Comment (bjorn) [inline src/libtiled/hex.h]: Please add a comment like:
```c++
/**
 * Represents a hex position in cube coordinates.
 */
```

File: src/libtiled/tilelayer.cpp (hunk 5 above)
Comment (Bdtrotte) [inline src/libtiled/tilelayer.cpp]: This was admittedly a bit of a lazy solution. Though I believe we don't have the inherent ability to change the stagger index from a TileLayer. The solution could be to omit the staggerIndex and staggerAxis parameters, and in place pass the map reference. I was thinking it would be preferable to keep them separate,  but I'll go ahead and implement that way if you think I'd be better.

File: src/libtiled/tilelayer.cpp (hunk 7 above)
Comment (Bdtrotte) [inline src/libtiled/tilelayer.cpp]: If a rectangle is rotated 60 degrees clockwise around the center, then the top right corner will be the furthest right point, the bottom left will be the furthest left by the same amount, and top left will be the highest, bottom right the lowest. Similar logic applies for anti-clockwise rotation, and the sizes will be the same, so one direction can be chosen, regardless of intended rotation direction. So we take the two corners, transform them around the center, rotate, and take the respective largest axis. This value is of the center, so it is doubled to get the maximum new width/height. The two is added because of potential integer rounding in regards to the center. 

Of course a problem is that sometimes the corners will be empty, so the enlargement is unnecessary. I don't believe there is an easy way to account for this, thus I enlarge, then trim it down later.

In short: This process produces the smallest rectangle which contains the original rotated by 60 degrees.

File: src/libtiled/tilelayer.h [hunk 11]
```diff
@@ -266,7 +267,7 @@ class TILEDSHARED_EXPORT TileLayer : public Layer
      * are rotated within the layer, and the tiles themselves are rotated. The
      * dimensions of the tile layer are swapped.
```
Comment (bjorn) [inline src/libtiled/tilelayer.h]: Please still adjust this comment, and make a note of the `map` parameter and that the stagger axis may be changed.

**Comment by bjorn** [comment]
Many thanks for this great enhancement!

**Comment by bjorn** [comment]
@Bdtrotte Please read up on Git, in particular how to update your fork without deleting and re-creating it. :-)

**Comment by Bdtrotte** [comment]
Of course! Will do. It was just pretty messy, so what quicker way to clean up..!
